src/meta_schedule/postproc/rewrite_tensorize.cc

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <tvm/runtime/container/base.h>
+
+#include <algorithm>
+
+#include "../utils.h"
+
+namespace tvm {
+namespace meta_schedule {
+
+using tir::BlockRV;
+using tir::LoopRV;
+
+void ApplyTensorization(const tir::Schedule& sch, const String& func_name,
+                        const tir::PrimFuncNode* func, bool vectorize_init_loop) {
+  std::vector<std::pair<std::string, std::function<void(tir::BlockRV)>>> jobs;
+
+  tir::PostOrderVisit(func->body, [=, &jobs](const ObjectRef& obj) {
+    if (const auto* block = obj.as<tir::BlockNode>()) {
+      tir::StmtSRef block_sref = sch->GetSRef(block);
+      if (Optional<String> intrin_name =
+              tir::GetAnn<String>(block_sref, tir::attr::meta_schedule_auto_tensorize)) {
+        std::string block_name = block_sref->StmtAs<tir::BlockNode>()->name_hint;
+        if (block_name.find("init") == std::string::npos) {
+          jobs.emplace_back(block_name, [sch, intrin_name](tir::BlockRV block) {
+            try {
+              sch->Tensorize(block, intrin_name.value());
+            } catch (const std::exception& e) {
+              LOG(WARNING) << "Tensorize failed with error " << e.what();
+            }
+          });
+        } else if (vectorize_init_loop) {
+          jobs.emplace_back(block_name, [sch](tir::BlockRV block) {
+            Array<BlockRV> child_blocks = sch->GetChildBlocks(block);
+            ICHECK(child_blocks.size() == 1);
+            Array<LoopRV> init_loops = sch->GetLoops(child_blocks[0]);
+            ICHECK(init_loops.size() == 1);
+            sch->Vectorize(init_loops[0]);
```

Comment:
Related to above, since `DecomposeReduction` introduces a new loop that should be vectorized on CPU, for now I'm applying vecotorization to the decomposed init loop here. This can also be done in `RewriteReductionBlock`.

Reply:
does `postproc::RewriteParallelVectorizeUnroll` for this case?

Reply:
I hope it would, but it doesn't. Also since parallelization etc is supposed to be applied before `DecomposeReduction`, I don't think running `RewriteParallelVectorizeUnroll` after `RewriteReductionBlock()` is a good idea. So vectorization of the init loop has to be done manually somehow.

I'd prefer vectoring in the init loop right after we run `DecomposeReduction`, since vecotorization of the init loop should be done on CPU regardless of tensorization. cc @MasterJH5574

Reply:
Interesting! What’s the order of post-processors being applied now? Perhaps we should reflect this order by adding this post-processor to tune.py https://github.com/apache/tvm/blob/effc23df7cb4c392f6957c94157da33293793eb7/python/tvm/meta_schedule/tune.py#L159-L170

Reply:
The issue in question is vectorization for CPU targets. I'm using the default postprocs in https://github.com/apache/tvm/blob/effc23df7cb4c392f6957c94157da33293793eb7/python/tvm/meta_schedule/tune.py#L96-L103

Since loop parallelization or vectorization checks for the "compact block" constraint, https://github.com/apache/tvm/blob/0ddaaa6a7d1009ea7ca8313a51eb19abb8ae7699/src/tir/schedule/primitive/for_kind.cc#L160, they need to be applied before `DecomposeReduction` in `RewriteReductionBlock()`. So having `RewriteParallelVectorizeUnroll` before `RewriteReductionBlock()` in the default postprocs makes sense.

However, this is not sufficient to vectorize the init loop of reduction block, since it is generated during `RewriteReductionBlock()`. I don't think we should run `RewriteReductionBlock()` again (and it doesn't work anyway), so we need to manually vectorize the decomposed init loop in `RewriteReductionBlock` or the new `RewriteTensorize` postproc I added. I prefer the former.

Reply:
I see. Thanks for the explanation. Just curious, is there any problem if we put `RewriteTensorize` before `RewriteParallelVectorizeUnroll`?

Reply:
`RewriteTensorize` needs `DecomposeReduction` to be applied beforehand, so it needs to be put after `RewriteReductionBlock`, which in turn follows `RewriteParallelVectorizeUnroll`. So the ordering is fixed.

Reply:
Okay the order makes sense to me now. But I still have another question 🤦‍♂️. Doesn’t `RewriteReductionBlock` vectorize the innermost loop outside the decomposed init block? I remember that the `DecomposeReduction` primitive invoked by the post-processor would indeed copy the loop kind https://github.com/apache/tvm/blob/60e43e16a0cfd7a215f94dadde03f197b12af3f4/src/tir/schedule/primitive/reduction.cc#L298

Reply:
In this case I want to tensorize the reduction block. So before `DecomposeReduction` is called, the loop kind  of the reduction is `serial`, which makes the decomposed init loop be `serial` as well.

Reply:
I see. So the block we want to tensorize wasn’t applied by the schedule rule `ParallelVectorizeUnroll` as well 🤔?

Reply:
ah yes, I'm not exactly sure what prevents `ParallelVectorizeUnroll` from tampering the block we want to tensorize (which is a good thing), maybe `Blockize` I do at 
https://github.com/apache/tvm/blob/dba2b31ce2a108b221c282f015b0d7fe09ea9358/src/meta_schedule/schedule_rule/multi_level_tiling_with_intrin.cc#L34
(after tiling the inner loop nests to be tensorized) is helping?

Reply:
Quite interesting.. So here the case is, on one hand we don’t want the block being annotated by rule `ParallelVectorizeUnroll`, but on the other hand we do want its init block to be vectorized after the decomposition. Am I right?

Since before decomposition the block wasn’t annotated by `ParallelVectorizeUnroll`, the decomposed init block isn’t vectorized, which makes sense. In addition, the decomposed init block doesn’t have any information to indicate that it’s supposed to vectorized (e.g., it doesn’t have an “need vectorization” annotation). In this case, no matter we vectorize the init block loop in `RewriteReductionBlock` or `RewriteTensorize`, it’s all due to our human knowledge, which I don’t think is perfect.

For upstreaming, it might be okay to do manual vectorization in `RewriteTensorize` (how does this vectorization bypass the compact dataflow issue BTW?). But in the long term I suppose we should enhance the compact dataflow check to allow such vectorization. After all, such vectorization won’t incur any incorrectness.

cc @junrushao1994 @spectrometerHBH

Reply:
> Quite interesting.. So here the case is, on one hand we don’t want the block being annotated by rule ParallelVectorizeUnroll, but on the other hand we do want its init block to be vectorized after the decomposition. Am I right?

Exactly.

> how does the vectorization in RewriteTensorize bypass the compact dataflow issue BTW?

That's a great question! Until recently, vectorization of the init loop after `DecomposeReduction` was rejected by the compact dataflow check. I brought this topic to @Hzfengsy and the team came up with a relaxation of the constraint that allows vectorizing init loop. This is the PR https://github.com/apache/tvm/pull/10705

Yeah, the ideally all outer loop parallelizations and inner loop vectorization can be done by one pass of `ParallelVectorizeUnroll`, meaning we run it after `DecomposeReduction`. Currently outer loop parallelization after `DecomposeReduction would be rejected by the compact dataflow check, but I think this is too restrictive.